src/Microsoft.IdentityModel.Logging/LogHelper.cs

```diff
@@ -37,6 +38,36 @@ namespace Microsoft.IdentityModel.Logging
     /// </summary>
     public class LogHelper
     {
+        /// <summary>
+        /// Gets or sets a logger to which logs will be written to.
+        /// </summary>
+        public static IIdentityLogger Logger { get; set; } = NullIdentityModelLogger.Instance;
```

Comment:
This being a static will cause problems down the road. Much like the JSON serializer and the claim type mapping setting.

Reply:
Could you expand on this or provide any relevant links on the issue?

Reply:
It's hard to find specific issues for 8+ years of history. I think there are relevant, but unless you lived thru those issues it's hard to explain.

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/202

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/550#issuecomment-267518199

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1475

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1465

https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1580

Reply:
@brockallen @sruke @AndersAbel the issues are semi-related to statics but seem mostly related to the general dislike of the decision to map claims and internal newton soft.

Our goal here is to provide a natural way for plugging in asp.net ILogger so that all logs are piped to the same listener.
We also want to expose this in such a way that layers, other than asp.net can plug so that new SDK's (we have in the pipeline) will have a unified logging model based on ILogger.

Any ideas are welcome.

I will look this PR over.

Reply:
@AndersAbel can you help us understand clearly by an example?

Reply:
> Any ideas are welcome.

Perhaps have a per-instance logger that's optional, and then fall back to the static/shared instance.

Reply:
@brockallen LogHelper is essentially a static class i am not sure how a per-instance would work.
Are you thinking that it would be helpful to have two instances of JwtSecurityTokenHandler use two different instances of LogHelper or a perhaps a different ILogger?

Reply:
I've not read all of this code, nor do I know the design. I assumed maybe there was an options instance somewhere and maybe that'd be where you'd hang a per-instance logger. But if not, then maybe there's nothing to be done.

Reply:
As a library author using MIM, I am *very* careful to not alter any global state/config in MIM, which would be the case with a static property. Consider this case:

```
App ----------------- Microsoft.IdentityModel
  |                      /
  |                     /
  +------Library-------/
```

An app references Microsoft.IdentityModel both directly and indirectly through a Library that has a dependency. For any config/setting that is per-instance, the library can create an own instance of the MIM object and do whatever it wants without affecting the behaviour seen by the applications direct usage. But if the property is static there are really just two bad options:

1. State in library documentation that application needs to do certain settings (i.e. assign a logger) to get intended functionality.
2. Set the static property from the Library and thus make the direct reference (that might have existed before) behave differently for the App.

Reply:
@AndersAbel i agree with your analysis about statics.

Currently all the objects in our runtime use a single instance of the current logging object.
There is no way to have two instances of JsonWebTokenHandler use a different logger or even with different PII setting.
As @brockallen suggested, we could develop a model where our TokenHandlers by default use the 'global' but will accept an instance of 'ILogger'.

What this work is about is developing a way for users to tell M.IM: "use this logger".
In this way, we hope that users can have an easy way to pipe all logs to a single location through multiple layers.
We can supply an adaptor that allows easy usage of asp.net ILogger.